pkg/transform/transform.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform
+
+import (
+	"context"
+	"fmt"
+
+	registryv1alpha1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/registry/v1alpha1"
+	"github.com/bufbuild/buf/private/pkg/protodescriptor"
+	"github.com/bufbuild/buf/private/pkg/protoencoding"
+	"github.com/bufbuild/connect-go"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+// ConvertMessage allows the caller to convert a given message data blob from
+// one format to another by referring to a type schema for the blob.
+func (c *Client) ConvertMessage(ctx context.Context, messageName string, inputData []byte) ([]byte, error) {
+	if len(c.types) > 0 && !contains(c.types, messageName) {
+		return nil, fmt.Errorf("message_name '%s' is not found in filtered types %v", messageName, c.types)
+	}
+	// First step: get descriptors for the requested message
+	resolver, err := c.getProtoEncodingResolver(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Now we can do the conversion.
+	md, err := resolver.FindMessageByName(protoreflect.FullName(messageName))
+	if err != nil {
+		return nil, err
+	}
+	msg := dynamicpb.NewMessage(md.Descriptor())
+
+	unmarshaler := getUnmarshaler(c.inputFormat, resolver, c.discardUnknown)
+	if err := unmarshaler.Unmarshal(inputData, msg); err != nil {
+		return nil, fmt.Errorf("input_data cannot be unmarshaled to %s in %s: %w", messageName, c.inputFormat, err)
+	}
+	marshaller, err := getMarshaller(c.outputFormat, resolver)
+	if err != nil {
+		return nil, err
+	}
+	data, err := marshaller.Marshal(msg)
+	if err != nil {
+		return nil, fmt.Errorf("input message cannot be unmarshaled to %s: %w", c.inputFormat, err)
+	}
+	// Done!
+	return data, nil
+}
+
+// getProtoEncodingResolver allows the caller to download a schema for one or more requested
+// types, RPC services, or RPC methods.
+func (c *Client) getProtoEncodingResolver(ctx context.Context) (protoencoding.Resolver, error) {
+	if c.cache != nil {
```

Comment:
I'm thinking maybe we _always_ use an in-memory cache, that is loaded/updated synchronously. And then we only fire off a goroutine to asynchronously save the results. We really only need to load the schema from remote source at startup (when there's nothing in memory for it).

We can't efficiently store a resolver into redis, for example: we'd have to serialize it as a file descriptor set. If only using redis (and not in-memory), this could be extremely inefficient if there are many messages to convert, because it would be constantly re-downloading the schema from redis and reconstructing the resolver.